### ISSUE TITLE
Be smarter about multiple concurrent iD sessions on the same site

### ISSUE THREAD

**Issue by quincylvania** [openstreetmap/iD#7250]
It's currently possible to have iD open in multiple tabs or windows at once. However, this use case is unexpected and is very likely to cause issues. Only one session can save edits locally, so data could easily be lost. Multiple sessions can change the same preferences, causing inconsistent states. We should prevent users from starting multiple sessions at all.

It's still fine to have iD open on different browsers (e.g. Safari and Chrome) or instances (e.g. `openstreetmap.org/edit` and `preview.ideditor.com/release`) since they aren't sharing the same local storage.

**Comment by thibaultmol** [comment]
(Might just be me) but I have situations where I open a second iD tab to view something from osmose, or some other qa tool, or even an id editor link from some chat room if someone has a question and I want to take a look.

**Comment by quincylvania** [comment]
@thibaultmol Yep, I agree iD is useful as a viewer like that. But it's still unsafe to start multiple edit sessions right now. I'd recommend saving your existing edits before viewing the link or using a different browser.

**Comment by thibaultmol** [comment]
I guess maybe having a "do you want to discard your other session" or "continue without auto save"? or something like that

**Comment by quincylvania** [comment]
> I guess maybe having a "do you want to discard your other session" or "continue without auto save"? or something like that

@thibaultmol Yeah, I've seen web apps that say something like "You have another open tab, only one is supported. **Switch to this one.**"

**Comment by thibaultmol** [comment]
Yeah, that'd work. But having the option to dismiss that popup and just
continue anyway would be ideal.
(Maybe adding a red text box in the to right to say something like "auto
save not available on this tab" so people still know which one is actually
being auto saved)

**Comment by quincylvania** [comment]
> But having the option to dismiss that popup and just
> continue anyway would be ideal.
> (Maybe adding a red text box in the to right to say something like "auto
save not available on this tab" so people still know which one is actually
being auto saved)

Hmm the two real options I see are to have multiple sessions work flawlessly (which would be needlessly difficult if not impossible) or not work at all. We don't have enough resources to support every kind of workflow, unfortunately.

**Comment by Thorwynn** [comment]
Would it be possible to have subsequent tabs open in read-only mode so objects & tags can be examined but not changed?

**Comment by quincylvania** [comment]
> Would it be possible to have subsequent tabs open in read-only mode so objects & tags can be examined but not changed?

This _sounds_ nice enough but would be a lot of work and could still result in users creating inconsistent settings between tabs that result in all kinds of difficult bugs. There's no simple compromise here. I don't think the vast majority of iD users would have any problem just uploading their current edits before going to view something else.

**Comment by 1ec5** [comment]
Is this a difference between v2 and v3, or has it been just as much of an issue forever? This limitation makes sense but comes as a surprise to me, since I’ve often had multiple tabs with iD open while editing without apparent problems.

Aside from locally saved changes, what’s the practical downside if preferences get inconsistent between tabs? Does all the code that depends on a preference read from local storage directly, or is there a layer of indirection that can cache values (which would at least ensure internal consistency)?

**Comment by 1ec5** [comment]
> Aside from locally saved changes, what’s the practical downside if preferences change out from under iD in local storage? Does all the code that depends on a preference read from local storage directly, or is there a layer of indirection that can cache values (which would at least ensure internal consistency)?

If this is only about history in local storage getting clobbered, would we be able to store per-tab session history by implementing #3239 or, even simpler, by saving multiple local storage entries (with, say, `_1` appended to the key) and remembering which one the current session is writing to?

**Comment by quincylvania** [comment]
@1ec5 At this point I agree that iD should be smarter about multiple concurrent sessions and not just prevent them altogether. There are use cases for wanting a lot of sessions open at once, but there is still the risk of data loss.
